src/java.base/windows/native/libnet/NetworkInterface.c

```diff
@@ -789,366 +286,389 @@ JNIEXPORT jobject JNICALL Java_java_net_NetworkInterface_getByName0
+        return JNI_FALSE;
+    }
+    return ifRow.Type == IF_TYPE_SOFTWARE_LOOPBACK ? JNI_TRUE : JNI_FALSE;
 }
 
 /*
  * Class:     java_net_NetworkInterface
  * Method:    getMacAddr0
  * Signature: ([bLjava/lang/String;I)[b
  */
-JNIEXPORT jbyteArray JNICALL Java_java_net_NetworkInterface_getMacAddr0
-    (JNIEnv *env, jclass class, jbyteArray addrArray, jstring name, jint index) {
-  jbyteArray ret = NULL;
-  int len;
-  MIB_IFROW *ifRowP;
-
-  // Retained for now to support IPv4 only stack, java.net.preferIPv4Stack
-  if (ipv6_available()) {
-    return Java_java_net_NetworkInterface_getMacAddr0_XP(env, class, name, index);
-  } else {
-    ifRowP = getIF(index);
-    if (ifRowP != NULL) {
-      switch(ifRowP->dwType) {
-      case MIB_IF_TYPE_ETHERNET:
-      case MIB_IF_TYPE_TOKENRING:
-      case MIB_IF_TYPE_FDDI:
-      case IF_TYPE_IEEE80211:
-        len = ifRowP->dwPhysAddrLen;
-        if (len > 0) {
-            ret = (*env)->NewByteArray(env, len);
-            if (!IS_NULL(ret)) {
-              (*env)->SetByteArrayRegion(env, ret, 0, len, (jbyte *) ifRowP->bPhysAddr);
-            }
-        }
-        break;
-      }
-      free(ifRowP);
+JNIEXPORT jbyteArray JNICALL Java_java_net_NetworkInterface_getMacAddr0(JNIEnv *env, jclass class, jbyteArray addrArray, jstring name, jint index) {
+    MIB_IF_ROW2 ifRow = {0};
+    jbyteArray macAddr;
+
+    ifRow.InterfaceIndex = index;
+    if (GetIfEntry2(&ifRow) != NO_ERROR) {
+        return NULL;
+    }
+    if (ifRow.PhysicalAddressLength == 0) {
+        return NULL;
     }
-    return ret;
-  }
+    macAddr = (*env)->NewByteArray(env, ifRow.PhysicalAddressLength);
+    if (macAddr == NULL) {
+        return NULL;
+    }
+    (*env)->SetByteArrayRegion(env, macAddr, 0, ifRow.PhysicalAddressLength, (jbyte *) ifRow.PhysicalAddress);
+    return macAddr;
 }
 
 /*
  * Class:       java_net_NetworkInterface
  * Method:      getMTU0
  * Signature:   ([bLjava/lang/String;I)I
  */
-JNIEXPORT jint JNICALL Java_java_net_NetworkInterface_getMTU0
-    (JNIEnv *env, jclass class, jstring name, jint index) {
-  jint ret = -1;
-  MIB_IFROW *ifRowP;
-
-  // Retained for now to support IPv4 only stack, java.net.preferIPv4Stack
-  if (ipv6_available()) {
-    return Java_java_net_NetworkInterface_getMTU0_XP(env, class, name, index);
-  } else {
-    ifRowP = getIF(index);
-    if (ifRowP != NULL) {
-      ret = ifRowP->dwMtu;
-      free(ifRowP);
+JNIEXPORT jint JNICALL Java_java_net_NetworkInterface_getMTU0(JNIEnv *env, jclass class, jstring name, jint index) {
+    MIB_IF_ROW2 ifRow = {0};
+
+    ifRow.InterfaceIndex = index;
+    if (GetIfEntry2(&ifRow) != NO_ERROR) {
+        return -1;
+    }
+    return ifRow.Mtu;
+}
+
+/*
+ * Class:     java_net_NetworkInterface
+ * Method:    supportsMulticast0
+ * Signature: (Ljava/lang/String;I)Z
+ */
+JNIEXPORT jboolean JNICALL Java_java_net_NetworkInterface_supportsMulticast0(JNIEnv *env, jclass cls, jstring name, jint index) {
```

Comment:
I'm not a big fan of using WBEM here. Also, I'm still trying to figure out when (if ever) this method is supposed to return false

Reply:
You and me both. As I wrote in the giant comment, I did a lot of experimenting. In the end, what I determined was that if the IGMPLevel (IPv4) or MldLevel (IPv6) is set to 0 (None), then attempting to join a multicast group throws an exception. So that seemed like a reasonable thing that supportsMulticast() could measure. Nothing else I tried, including the existing GetAdaptersAddresses approach, indicated that multicast was disabled in those situations.

Having said that, I'd be okay with ditching all of this code and just returning true unconditionally. It's unlikely that multicast would be disabled, and a single boolean isn't enough to express the actual setting anyway, since Windows apparently lets you disable receiving multicasts while still allowing you to send them.

Reply:
Ha! Apparently I have a machine where the implementation of `supportsMulticast` implementation makes a difference. On my machine the test `jdk/java/net/SocketOption/OptionsTest.java` is failing with:
```
java.net.SocketException: Invalid argument: no further information
	at java.base/sun.nio.ch.Net.setInterface6(Native Method)
	at java.base/sun.nio.ch.DatagramChannelImpl.setOption(DatagramChannelImpl.java:377)
	at java.base/sun.nio.ch.DatagramSocketAdaptor.setOption(DatagramSocketAdaptor.java:431)
	at java.base/java.net.DatagramSocket.setOption(DatagramSocket.java:1207)
	at OptionsTest.test(OptionsTest.java:155)
	at OptionsTest.doMulticastSocketTests(OptionsTest.java:235)
	at OptionsTest.main(OptionsTest.java:337)
...
```
It works with the current master. Will try to figure it out.

BTW. did you try running jdk_net tests?

Reply:
I ran the tests for NetworkInterface specifically, though that reminds me that I need to open a bug report about fixpath mangling things incorrectly. I'll give the full jdk_net suite a run and see what happens.

Reply:
ok so here goes: your patch changes the order in which the interfaces are returned; the original code returned them in order of increasing indexes, the new code appears to sort interfaces by LUID.
The failing test uses the first interface for which supportsMulticast returns true. I get `supportsMulticast` = `true` on all interfaces even with the old code.

The interface that happened to be on the first position of my list was:
```
67 name:ethernet_0 (Hyper-V Virtual Switch Extension Adapter-Hyper-V Virtual Switch Extension Filter-0000)
 ifup: true isP2P: false isLoopback: false macAddr: null MTU: 1500 supportsMulticast: true
```
it has no associated IP addresses, and is not usable with IP_MULTICAST_IF.
(Also it's a filter interface, which means that the old code returned ifUp = false).

Well that would explain how `supportsMulticast` is meant to be used; now we just need to figure out which interfaces can be used for multicasting.

Reply:
Note that the new code doesn't attempt to sort anything. If it happens to be sorted by LUID, that's the underlying API at work. The MS docs don't say anything about ordering, so this is probably an implementation detail and I wouldn't assume it will always be true.

I'm still running jdk_net, but I can already see that test failing on my machine. Though it has a different error:

```
java.io.IOException: Network interface not configured for IPv4
	at java.base/sun.nio.ch.DatagramChannelImpl.setOption(DatagramChannelImpl.java:389)
	at java.base/sun.nio.ch.DatagramSocketAdaptor.setOption(DatagramSocketAdaptor.java:431)
	at java.base/java.net.DatagramSocket.setOption(DatagramSocket.java:1207)
	at OptionsTest.test(OptionsTest.java:155)
	at OptionsTest.doMulticastSocketTests(OptionsTest.java:235)
	at OptionsTest.main(OptionsTest.java:337)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
	at java.base/java.lang.Thread.run(Thread.java:1623)
```

So I think when picking the interface to use for the test, we need to also check that isUp = true and that the list of IP addresses contains at least one IPv4 address.

Reply:
Though it's more complicated than that. Some machines may have both IPv6 and IPv4, some machine may only have IPv4, and some machine may only have IPv6. I haven't looked at the test yet, but picking up the right interface on the various platforms has always been an issue for our tests.

Reply:
So,, apparently this is a Windows feature. [Older socket implementation](https://github.com/openjdk/jdk11/blob/37115c8ea4aff13a8148ee2b8832b20888a5d880/src/java.base/windows/native/libnet/TwoStacksPlainDatagramSocketImpl.c#L1350) had an explicit check for addresses bound to an interface, and only interfaces with associated addresses could be used for multicasting.

This needs to be fixed in the test. Can you add a check for addresses to `OptionsTest.getNetworkInterface`?

Beyond that, I'm fine with an implementation that unconditionally returns `true`.

Reply:
I've adjusted OptionsTest as @djelinski suggested, and changed the implementation to always return true. I left the giant comment so that when future devs wonder WTF we were thinking, they'll have the answer.